### PR TITLE
Speed up the peft lora unload

### PR DESCRIPTION
**Pull request by pacman100**
# What does this PR do?
1. Speed up the peft lora unload
2. PR https://github.com/huggingface/peft/pull/1106 should be merged first

### CODE REVIEW

File: src/diffusers/utils/peft_utils.py [hunk 1]
```diff
@@ -24,54 +24,72 @@
...
-            setattr(model, name, new_module)
-            del module
-
-            if torch.cuda.is_available():
-                torch.cuda.empty_cache()
+    from peft.tuners.tuners_utils import BaseTunerLayer
 
+    for module in model.modules():
+        if isinstance(module, BaseTunerLayer):
+            has_base_layer_pattern = True if hasattr(module, "base_layer") else False
```
Comment (BenjaminBossan) [inline src/diffusers/utils/peft_utils.py:34]:
```suggestion
            has_base_layer_pattern = hasattr(module, "base_layer")
```

File: src/diffusers/utils/peft_utils.py [hunk 2]
```diff
@@ -24,54 +24,72 @@
...
+        for key in key_list:
+            try:
+                parent, target, target_name = _get_submodules(model, key)
+            except AttributeError:
+                continue
+            if hasattr(target, "base_layer"):
+                setattr(parent, target_name, target.get_base_layer())
+    else:
+        if is_torch_available():
+            import torch
```
Comment (BenjaminBossan) [inline src/diffusers/utils/peft_utils.py:50]: I know this was already there before, but if torch is not available, the code below will just error, right?

File: src/diffusers/utils/peft_utils.py [hunk 3]
```diff
@@ -24,54 +24,72 @@
...
+
+        key_list = [key for key, _ in model.named_modules() if "lora" not in key]
+        for key in key_list:
+            try:
+                parent, target, target_name = _get_submodules(model, key)
+            except AttributeError:
+                continue
+            if hasattr(target, "base_layer"):
+                setattr(parent, target_name, target.get_base_layer())
+    else:
```
Comment (BenjaminBossan) [inline src/diffusers/utils/peft_utils.py:48]: Maybe add a comment like: "This is for backwards compatibility with PEFT <= 0.6.2. TODO can be removed once that PEFT version is no longer supported."

Not sure if diffusers has a specific way to comment possible deprecations, I couldn't find anything.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks Sourab for making this work with the new PR on PEFT. I have some minor comments, no blockers.

For context, the mentioned PR is now merged and this change is necessary to work with PEFT main. We ran a test on OPT-350m, where we saw 15x speedup in unloading with the new changes.

File: src/diffusers/utils/peft_utils.py (hunk 3 above)
Comment (pacman100) [reply src/diffusers/utils/peft_utils.py:49]: Done

File: src/diffusers/utils/peft_utils.py (hunk 2 above)
Comment (younesbelkada) [reply src/diffusers/utils/peft_utils.py:53]: Correct yes, you can put torch on the top level of the file I think to mitigate that

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for this and for taking care of the backward compatibility! left one suggestion above

File: src/diffusers/utils/peft_utils.py (hunk 2 above)
Comment (pacman100) [reply src/diffusers/utils/peft_utils.py:53]: But that resulted in import errors

Comment (younesbelkada) [reply src/diffusers/utils/peft_utils.py:53]: even if you protect the import with `is_torch_available()`?

Comment (pacman100) [reply src/diffusers/utils/peft_utils.py]: done

**Review by BenjaminBossan — APPROVED** [review]
LGTM, thanks.

**Comment by patrickvonplaten** [comment]
Thanks!

**Pull request merged**